File: src/main/java/org/example/iotserver/controllers/BuildingController.java
package org.example.iotserver.controllers;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import org.example.iotserver.services.BuildingService;
import org.example.iotserver.models.Building;

@RestController
@RequestMapping("/api/buildings")
public class BuildingController {
    private final BuildingService buildingService;

    public BuildingController(BuildingService buildingService) {
        this.buildingService = buildingService;
    }

    // GET: Fetch all buildings
    @GetMapping
    public List<Building> getAllBuildings() {
        return buildingService.getAllBuildings();
    }


    @PostMapping
    public ResponseEntity<Building> createBuilding(@RequestBody Building building) {
        Building savedBuilding = buildingService.createBuilding(building);
        return ResponseEntity.status(HttpStatus.CREATED).body(savedBuilding);
    }

    // PUT: Update an existing building
    @PutMapping("/{id}")
    public Building updateBuilding(@PathVariable Long id, @RequestBody Building updatedBuilding) {
        return buildingService.updateBuilding(id, updatedBuilding);
    }

    // DELETE: Delete a building by ID
    @DeleteMapping("/{id}")
    public void deleteBuilding(@PathVariable Long id) {
        buildingService.deleteBuilding(id);
    }
}

File: backend/controllers/buildingController.js
const buildingService = require('../services/buildingService');

// Controller for fetching all buildings
exports.getAllBuildings = async (req, res) => {
  try {
    const buildings = await buildingService.getAll(); // Using the service layer
    res.status(200).json(buildings);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

// Controller for creating a new building
exports.createBuilding = async (req, res) => {
  try {
    const { name, address } = req.body;
    const building = await buildingService.create({ name, address }); // Using the service layer
    res.status(201).json(building);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Controller for updating a building
exports.updateBuilding = async (req, res) => {
  try {
    const { id } = req.params;
    const { name, address } = req.body;
    const building = await buildingService.update(id, { name, address }); // Using the service layer
    res.status(200).json(building);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
};

// Controller for deleting a building
exports.deleteBuilding = async (req, res) => {
  try {
    const { id } = req.params;
    await buildingService.delete(id); // Using the service layer
    res.status(204).send();
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};
